File: Elixir/process_spec.ex
defmodule ProcessSpec do
  def spawn_client(reply_time) do
    {:ok, spawn_link(__MODULE__, :client, [reply_time])}
  end

  def client(reply_time) do
    receive do
      {"Hello", pid, data} ->
        Process.sleep(reply_time)
        send(pid, {"Done", self(), data})
        client(reply_time)
      _ ->
        IO.puts("Unexpected message received!")
    end
  end

  def spawn_server(counter, data, reply_time) do
    pid = spawn_link(__MODULE__, :server, [counter, spawn_link(__MODULE__, :client, [reply_time]), 0, data])
    send(pid, "Start")
    {:ok, pid}
  end

  def server(counter, clients, message_count, data) do
    receive do
      "Start" ->
        send(clients, {"Hello", self(), data})
        server(counter, clients, message_count, data)
      {"Done", pid, data} ->
        send(counter, 1)
        send(pid, {"Hello", self(), data})
        server(counter, clients, message_count+1, data)
      "Stop" ->
        exit(:kill)
      _ ->
        IO.puts("Unexpected message received!")
    end
  end
  def spawn_counter(interval, summation_mode) do
    pid = spawn_link(__MODULE__, :aggregator, [0, :erlang.system_time(:millisecond), [], summation_mode])
    Process.register(pid, :counter)
    if summation_mode != :avg_total do
      spawn_link(__MODULE__, :logger, [pid, interval])
    end
    {:ok, pid}
  end
  def aggregator(n, start_time, record, mode) do

    receive do
      "Start" ->
        aggregator(0, :erlang.system_time(:millisecond), [], mode)
      "Update" ->
        elapsed_time = (:erlang.system_time(:millisecond)-start_time)/1000
        aggregator(0, :erlang.system_time(:millisecond), [n/elapsed_time | record], mode)
      "Stop" ->
        {:ok, file} = File.open("r_400_500_ta_23.txt", [:append])
        case mode do
          :avg ->
            average = :lists.sum(record) / length(record)
            :io.format(file, "~.3f~n", [average])
          :full ->
            for n <- :lists.reverse(record) do
              :io.format(file, "~.2f~n", [n])
            end
          :avg_total ->
            :io.format(file, "~.3f~n", [(n*1000)/(:erlang.system_time(:millisecond)-start_time)])
        end
      messages_count ->
        aggregator(n+messages_count, start_time, record, mode)
    end
  end

  def logger(counter, interval) do
    Process.sleep(interval)
    send(counter,"Update")
    logger(counter, interval)
  end
end

File: Erlang/processSpec.erl
-module(processSpec).
-export([spawnClient/1, spawnServer/3, spawnCounter/2, spawnLogger/1]).
-export([client/1, server/4, aggregator/4, logger/2]).

spawnClient(ReplyTime) ->
    {ok, spawn_link(?MODULE, client, [ReplyTime])}.

client(ReplyTime) ->
    receive
        {"Hello", Pid, Data} ->
            timer:sleep(ReplyTime),
            Pid ! {"Done", self(), Data},
            client(ReplyTime);
        _ ->
            io:format("Unexpected message received!~n")
    end.

spawnServer(Counter, Data, ReplyTime) ->
    Pid = spawn_link(?MODULE, server, [Counter, spawn_link(?MODULE, client, [ReplyTime]), 0, Data]),
    Pid ! "Start",
    {ok, Pid}.

server(Counter, Clients, MessageCount, Data) ->
    receive
        "Start" ->
            Clients ! {"Hello",self(), Data},
            server(Counter, Clients, MessageCount, Data);
        {"Done", Pid, Data} ->
            Counter ! 1,
            Pid ! {"Hello",self(), Data},
            server(Counter, Clients, MessageCount + 1, Data);
        "Stop" ->
            exit(kill);
        _ ->
            io:format("Unexpected message received!~n")
    end.

spawnCounter(Interval, SummationMode) ->
    Pid = spawn_link(?MODULE, aggregator, [0, erlang:system_time(millisecond), [], SummationMode]),
    register(counter, Pid),
    case SummationMode of
        totalavg-> ok ;
        _-> spawn_link(?MODULE, logger, [Pid, Interval])
    end,
    {ok, Pid}.

aggregator(N, StartTime, Record, SummationMode) ->
    receive
        "Start" ->
            aggregator(0, erlang:system_time(millisecond), [], SummationMode);
        "Update" ->
                ElapsedTime = (erlang:system_time(millisecond) - StartTime)/1000,
                Throughput = N/ElapsedTime,
                aggregator(0, erlang:system_time(millisecond), [Throughput | Record], SummationMode);
        "Stop" ->
            {ok, Fd} = file:open("result.txt", [append]),
            case SummationMode of
                full -> [io:format(Fd, "~.2f~n", [D]) || D <- lists:reverse(Record)];
                avg -> io:format(Fd, "~.3f~n", [lists:sum(Record) / length(Record)]);
                totalavg ->  io:format(Fd, "~.3f~n", [(N*1000)/(erlang:system_time(millisecond)-StartTime)])
            end,
            counter ! stop;
        MessagesCount ->
            aggregator(N+MessagesCount, StartTime, Record, SummationMode)
    end.

spawnLogger(Counter) ->
    {ok, spawn(?MODULE, logger, [Counter])}.

logger(Counter, Interval) ->
    timer:sleep(Interval),
    Counter ! "Update",
    logger(Counter, Interval).